include/revng/ADT/GenericGraph.h

```diff
@@ -353,11 +370,557 @@ class BidirectionalNode
+    Predecessors.emplace_back(std::move(View));
+    return EdgeView(Output);
+  }
+
+protected:
+  struct SuccessorFilters {
+    struct Types {
+      using ToView = EdgeView (*)(OwningEdge &);
+      using ToConstView = ConstEdgeView (*)(OwningEdge const &);
+      using ToNeighbor = DerivedType *&(*) (OwningEdge &);
+      using ToConstNeighbor =
+        DerivedType const *const &(*) (OwningEdge const &);
+    };
+
+    static constexpr typename Types::ToView ToView =
+      [](OwningEdge &E) -> EdgeView { return EdgeView(E); };
+    static constexpr typename Types::ToConstView ToConstView =
+      [](OwningEdge const &E) -> ConstEdgeView { return ConstEdgeView(E); };
+    static constexpr typename Types::ToNeighbor ToNeighbor =
+      [](OwningEdge &E) -> DerivedType *& { return E.Neighbor; };
+    static constexpr typename Types::ToConstNeighbor ToConstNeighbor =
+      [](OwningEdge const &E) -> DerivedType const *const & {
+      return E.Neighbor;
+    };
+  };
+
+  struct PredecessorFilters {
+    struct Types {
+      using ToView = EdgeView (*)(NonOwningEdge &);
+      using ToConstView = ConstEdgeView (*)(NonOwningEdge const &);
+      using ToNeighbor = DerivedType *&(*) (NonOwningEdge &);
+      using ToConstNeighbor =
+        DerivedType const *const &(*) (NonOwningEdge const &);
+    };
+
+    static constexpr typename Types::ToView ToView =
+      [](NonOwningEdge &E) -> EdgeView { return EdgeView(E); };
+    static constexpr typename Types::ToConstView ToConstView =
+      [](NonOwningEdge const &E) -> ConstEdgeView { return ConstEdgeView(E); };
+    static constexpr typename Types::ToNeighbor ToNeighbor =
+      [](NonOwningEdge &E) -> DerivedType *& { return E.Neighbor; };
+    static constexpr typename Types::ToConstNeighbor ToConstNeighbor =
+      [](NonOwningEdge const &E) -> DerivedType const *const & {
+      return E.Neighbor;
+    };
+  };
+
+  using SuccessorEdgeIterator = revng::
+    mapped_iterator<OwningEdge *, typename SuccessorFilters::Types::ToView>;
+  using ConstSuccessorEdgeIterator = revng::mapped_iterator<
+    OwningEdge *,
+    typename SuccessorFilters::Types::ToConstView>;
+  using SuccessorIterator = revng::
+    mapped_iterator<OwningEdge *, typename SuccessorFilters::Types::ToNeighbor>;
+  using ConstSuccessorIterator = revng::mapped_iterator<
+    OwningEdge *,
+    typename SuccessorFilters::Types::ToConstNeighbor>;
+
+  using PredecessorEdgeIterator = revng::mapped_iterator<
+    NonOwningEdge *,
+    typename PredecessorFilters::Types::ToView>;
+  using ConstPredecessorEdgeIterator = revng::mapped_iterator<
+    NonOwningEdge *,
+    typename PredecessorFilters::Types::ToConstView>;
+  using PredecessorIterator = revng::mapped_iterator<
+    NonOwningEdge *,
+    typename PredecessorFilters::Types::ToNeighbor>;
+  using ConstPredecessorIterator = revng::mapped_iterator<
+    NonOwningEdge *,
+    typename PredecessorFilters::Types::ToConstNeighbor>;
```

Comment:
I guess those were more for me, so that I don't get confused between all those iterator types. Being extra explicit often helps in the cases like that.
The problem is (not sure if it applies to all of them, but) their constructors are called. And the second parameter changes depending on exactly which iterator it is.

Moreover, some of them are also accepted as parameters.

Reply:
Yeah, I'm usually in favor of explicit but here's quite verbose, a bit confusing and error prone in copy pasting.
Code you don't write cannot be wrong :D

If you make something a bit smart like `revng::make_mapped_iterator` you should be able to avoid calling the constructors directly.

Reply:
I'll see what I can do.

Reply:
Ok, so I was able to avoid using the constructors directly, but I still can't exactly remove the definition of these types: they are used as input parameters of `remove(?:Successor|Predecessor)` and I'm not comfortable with making the `remove` stuff accepting any iterator; the more restrictions are in place, the lower the chance that a user can successfully pass into it something incompatible.

So, the question is, how strongly do you feel about these explicit iterator type declarations, because at this point I'm getting a nagging feeling that it's better to leave them be.

Reply:
Except those look even uglier (IMHO) after the conversion of lambdas into normal member functions: [code](https://github.com/revng/revng/blob/d1b08bdab6850a29eca4a8f520b845ee01847e23/include/revng/ADT/GenericGraph.h#L631).

Reply:
OK, let's leave it as it is.